mars/dataframe/groupby/apply.py

```diff
@@ -0,0 +1,213 @@
+                    [[applied.index[0]] * len(applied.columns), applied.columns.to_list()])
+                applied_idx.names = [applied.index.name, None]
+                applied = pd.Series(applied.iloc[0].to_numpy(), applied_idx, name=applied.columns.name)
+        else:
+            applied = grouped.transform(op.func, *op.args, **op.kwds)
+        ctx[op.outputs[0].key] = applied
+
+    @classmethod
+    def tile(cls, op):
+        in_groupby = op.inputs[0]
+        out_df = op.outputs[0]
+
+        chunks = []
+        for c in in_groupby.chunks:
+            inp_chunks = [c]
+
+            new_op = op.copy().reset_key()
+            if op.object_type == ObjectType.dataframe:
+                chunks.append(new_op.new_chunk(
+                    inp_chunks, index=c.index, shape=(np.nan, len(out_df.dtypes)), dtypes=out_df.dtypes,
+                    columns_value=out_df.columns_value, index_value=out_df.index_value))
+            else:
+                chunks.append(new_op.new_chunk(
+                    inp_chunks, index=c.index, shape=(np.nan,), dtype=out_df.dtype,
+                    index_value=out_df.index_value))
+
+        new_op = op.copy().reset_key()
+        kw = out_df.params.copy()
+        kw['chunks'] = chunks
+        if not op.is_transform:
+            if op.object_type == ObjectType.dataframe:
+                kw['nsplits'] = ((np.nan,) * len(chunks), (out_df.shape[1],))
+            else:
+                kw['nsplits'] = ((np.nan,) * len(chunks),)
+        return new_op.new_tileables([in_groupby], **kw)
+
+    def _infer_df_func_returns(self, in_object_type, in_dtypes, dtypes, index):
+        index_value, object_type, new_dtypes = None, None, None
+        if self.is_transform:
+            object_type = in_object_type
+
+        try:
+            if in_object_type == ObjectType.dataframe:
+                empty_df = build_empty_df(in_dtypes, index=pd.RangeIndex(2))
+            else:
+                empty_df = build_empty_series(in_dtypes, index=pd.RangeIndex(2))
+
+            with np.errstate(all='ignore'):
+                if self.is_transform:
+                    infer_df = empty_df.apply(self.func, *self.args, **self.kwds)
+                else:
+                    infer_df = self.func(empty_df, *self.args, **self.kwds)
+
+            if isinstance(infer_df, np.number):
+                index_value = parse_index(pd.RangeIndex(0, 1))
+            else:
+                # todo return proper index when sort=True is implemented
+                index_value = parse_index(pd.RangeIndex(-1))
+
+            if isinstance(infer_df, pd.DataFrame):
+                object_type = object_type or ObjectType.dataframe
+                new_dtypes = new_dtypes or infer_df.dtypes
+            elif isinstance(infer_df, pd.Series):
+                object_type = object_type or ObjectType.series
+                new_dtypes = new_dtypes or infer_df.dtype
+            else:
+                object_type = ObjectType.series
```

Comment:
`ObjectType` could be scalar which is actually a tensor.

Reply:
When the returned value of the provided function is a scalar, groupby().apply() actually returns a series.